### PR TITLE
Hero exchange dialog improvement

### PR DESCRIPTION
**Pull request by nullkiller**
The idea of this PR is to implement hero exchange dialog from HD mod. Now it contains only army exchange buttons working and I plan to implement artifact exchange buttons also. This thing can be merged now or we can wait for the rest of changes.

- [x] pick changes from @ars3niy (https://github.com/vcmi/vcmi/pull/636)
- [x] acquire resources
- [x] implement hero exchange
- [x] implement artifact exchange

### CODE REVIEW

File: client/windows/GUIClasses.cpp [hunk 1]
```diff
@@ -787,9 +787,95 @@ void CTavernWindow::HeroPortrait::hover(bool on)
...
+
+std::function<void()> CExchangeController::onMoveArmyToRight()
+{
+	return [&]() { hdModQuickExchangeArmy(true); };
+}
+
+std::function<void()> CExchangeController::onSwapArmy()
+{
+	return [&]() { 
+		auto cb = LOCPLINT->cb;
```
Comment (alexvins) [inline client/windows/GUIClasses.cpp:816]:
```suggestion
	return [&]() 
	{ 
		auto cb = LOCPLINT->cb;

```

File: client/windows/GUIClasses.cpp [hunk 2]
```diff
@@ -787,9 +787,95 @@ void CTavernWindow::HeroPortrait::hover(bool on)
...
+}
+
+void CExchangeController::hdModQuickExchangeArmy(bool leftToRight)
+{
+	auto source = leftToRight ? left : right;
+	auto target = leftToRight ? right : left;
+	auto cb = LOCPLINT->cb;
+
+	boost::thread([=]
+	{
```
Comment (alexvins) [inline client/windows/GUIClasses.cpp:833]: Please explain how threading helps here (but not in function above, both do server call for each slot).

**Review by alexvins — CHANGES_REQUESTED** [review]

File: client/windows/GUIClasses.cpp (hunk 2 above)
Comment (nullkiller) [reply client/windows/GUIClasses.cpp:833]: cb->waitTilRelaize locks main thread. For easier implementation here I need to see the results of cb invokations so I need to use waitTilRealize and thus I need a separate thread.

File: client/widgets/Images.cpp [hunk 3]
```diff
@@ -259,10 +290,26 @@ void CAnimImage::showAll(SDL_Surface * to)
 
 	if(flags & CShowableAnim::BASE && frame != 0)
 		if(auto img = anim->getImage(0, group))
-			img->draw(to, pos.x, pos.y);
+		{
```
[the remark this answers is not in the document]
Comment (nullkiller) [reply client/widgets/Images.cpp:293]: Sorry, I even have not seen it. Will take a closer look now

File: client/windows/GUIClasses.cpp (hunk 2 above)
Comment (alexvins) [reply client/windows/GUIClasses.cpp:833]: The question is why is a new thread here but in `CExchangeController::onSwapArmy()` for same amount of `cb` calls is not? Also if you spawn a thread you need (at least shared) lock of gs there, but this in turn will still lock UI. And with acquired lock you could use unlockGsWhenWaiting.

Comment (nullkiller) [reply client/windows/GUIClasses.cpp:833]: Because I do not need the result of those calls in swapArmy

Comment (nullkiller) [reply client/windows/GUIClasses.cpp:833]: Btw I think you spotted a nice bug if the first stack for swapping is the last one and the same slot of another army is empty

Comment (nullkiller) [reply client/windows/GUIClasses.cpp:835]: Is it ok now?

File: client/widgets/Images.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (nullkiller) [reply client/widgets/Images.cpp:301]: Fixed code duplication

File: client/windows/GUIClasses.cpp (hunk 2 above)
Comment (alexvins) [reply client/windows/GUIClasses.cpp:835]: Yes it is

**Review by alexvins — APPROVED** [review]

**Review by alexvins — CHANGES_REQUESTED** [review]

File: client/windows/GUIClasses.cpp [hunk 4]
```diff
@@ -787,9 +787,363 @@ void CTavernWindow::HeroPortrait::hover(bool on)
...
+		boost::thread(std::bind(&GsThread::run, std::make_shared<GsThreadState>(action)));
+	}
+
+private:
+	static void run(std::shared_ptr<GsThreadState> state)
+	{
+		boost::shared_lock<boost::shared_mutex> gsLock(CGameState::mutex);
+
+		auto originalWaitTillRealize = state->cb->waitTillRealize;
```
[the remark this answers is not in the document]
Comment (nullkiller) [reply client/windows/GUIClasses.cpp:824]: Makes sense

**Comment by nullkiller** [comment]
Refactored GsThread and now it is only left to update vcmi essentials or create a new separate hdmod. Anyway it works without HD mod resources, just uses old UI instead.

**Review by alexvins — APPROVED** [review]

**Pull request merged**